aesara/tensor/math_opt.py

```diff
@@ -3653,3 +3654,51 @@ def local_1msigmoid(fgraph, node):
 
 if register_local_1msigmoid:
     register_canonicalize(local_1msigmoid)
+
+
+# log(sigmoid(x) / (1 - sigmoid(x))) -> x
+# i.e logit(sigmoid(x)) -> x
+local_logit_sigmoid = PatternSub(
+    (log, (true_div, (sigmoid, "x"), (sub, 1, (sigmoid, "x")))),
+    "x",
+    allow_multiple_clients=True,
+    name="local_logit_sigmoid",
+)
+register_stabilize(local_logit_sigmoid)
+register_specialize(local_logit_sigmoid)
+
+# softplus(x) - softplus(-x) -> x
+# This form often arises when inverting the sigmoid graph. Ideally we would
+# write the pattern in the first line of the comment, but then aesara does not
+# recognize the more common reverse version -softplus(-x) - (-softplus(x))
+local_sub_softplus = PatternSub(
+    # (sub, (softplus, "x"), (softplus, (neg, "x"))),
+    (sub, (neg, (softplus, (neg, "x"))), (neg, (softplus, "x"))),
+    "x",
+    allow_multiple_clients=True,
+    name="local_sub_softplus",
+)
+register_stabilize(local_sub_softplus)
+register_specialize(local_sub_softplus)
+
+# sigmoid(log(x / (1-x)) -> x
+# i.e., sigmoid(logit(x)) -> x
+local_sigmoid_logit = PatternSub(
+    (sigmoid, (log, (true_div, "x", (sub, 1, "x")))),
+    "x",
+    allow_multiple_clients=True,
+    name="local_sigmoid_logit",
+)
+register_stabilize(local_sigmoid_logit)
+register_specialize(local_sigmoid_logit)
+
+# sigmoid(logit(x)) -> x, when log1p is present in the denominator
+# sigmoid(log(x / (1-x)) can easily get optimized to sigmoid(log(x) - log1p(-x))
```

Comment:
For canonicalization, something like log1p(x) should really be replaced with log(1 + x).

This seems to touch on the same issue described in https://github.com/aesara-devs/aesara/issues/441.

I agree with canonicalizing `log1p` to `log(1+x)` and later specializing / stabilizing with it. That however should be done in a separate PR as it will touch on many other previous rewrites.

> We can also consider specialized Ops like log1p to mean that the user is "forcing" it to be viewed as a canonical term.

I don't think we should take the explicit `log1p` to mean that a user does not want this graph to be temporarily canonicalized "away". That would be a very implicitly (and invisible) form of control, that is better addressed by manually disabling some of the default rewrites if they are not of interest.

Reply:
Also we should have a much more exhaustive documentation of which rewrites are applied and where if we want to give back control to users